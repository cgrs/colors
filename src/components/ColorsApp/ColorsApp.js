import React, { Component } from 'react'
import Color from 'color'
import './ColorsApp.css'
import ColorSelector from '../ColorSelector'

class ColorsApp extends Component {
  constructor () {
    super ()
    const hist = window.history.state || []
    this.state = {
      current: Color({h: 180, s: 50, l: 50}),
      colors: hist.map(Color)
    }
    this.currrentColorSelector = React.createRef()
    window.scrollTo({
      top: window.innerHeight/2
    })
  }

  mouseMove (e) {
    let {current} = this.state
    const {clientX, clientY} = e
    const rect = this.currrentColorSelector.current.getBoundingClientRect()
    const [x, y] = [
      clientX/rect.width,
      clientY/rect.height
    ]
    current = current
      .hue(Math.floor(360*x))
      .lightness(Math.floor(100*y))
    this.setState({...this.state, current})
  }

  saveColor () {
    const colors = this.state.colors.slice()
    colors.push(this.state.current)
    window.history.pushState(
      colors.map(c => c.hex()), document.title, colors.map(c => c.hex()).join()
    )
    this.setState({...this.state, colors})
  }
  changeSaturation (e) {
    let {current} = this.state
    const z = (window.scrollY*2) / document.body.scrollHeight
    current = current.saturationl(100*z)
    this.setState({...this.state, current})
  }
  render () {
    const colors = this.state.colors
    const current = this.state.current
    return (
      <div className='colors-app'>
        {
          colors.map(
            (color, index) => (
              <ColorSelector key={index} color={color} className='color-selector animated'/>
            )
          )
        }
        <ColorSelector
          color={current}
          className='color-selector'
          reference={this.currrentColorSelector}
          onMouseMove={e => this.mouseMove(e)}
          onClick={e => this.saveColor(e)}
          onWheel={e => this.changeSaturation(e)}
        />
      </div>
    )
  }
}

export default ColorsApp
