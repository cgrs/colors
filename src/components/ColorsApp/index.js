import ColorsApp from './ColorsApp'
export default ColorsApp
