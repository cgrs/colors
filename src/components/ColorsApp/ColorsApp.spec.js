import React from 'react'
import ColorsApp from './ColorsApp'
import renderer from 'react-test-renderer'

describe('ColorsApp', () => {
  it('renders correctly', () => {
    const component = renderer.create(<ColorsApp/>)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
