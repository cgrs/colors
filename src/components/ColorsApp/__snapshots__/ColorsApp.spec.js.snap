// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorsApp renders correctly 1`] = `
<div
  className="colors-app"
>
  <div
    className="color-selector"
    onClick={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "backgroundColor": "hsl(180, 50%, 50%)",
      }
    }
  >
    <div
      className="color-info info-container"
    >
      <p
        className="color-info color"
      >
        #40BFBF
      </p>
      <p
        className="color-info info"
      >
        click to save
      </p>
    </div>
  </div>
</div>
`;
